File: compilers/python/runar_compiler/frontend/side_effect_summary.py
"""Side-effect summary pass.

Mirrors ``packages/runar-compiler/src/passes/side-effect-summary.ts``.

Classifies each method on a ContractNode by the side effects it has on
the contract's continuation requirements. Walks the private-method
call graph so effects buried inside private helpers surface to their
public callers.

Consumed by ``anf_lower.py`` for:

  - Auto-injecting continuation parameters (``_changePKH``,
    ``_changeAmount``, ``_newAmount``, ``txPreimage``) on public
    stateful methods.
  - Gating emission of the hashOutputs continuation assertion.
  - Deciding whether a private-helper call should be inlined into the
    caller's binding stream so its add_output / add_data_output ANF
    nodes register on the caller's continuation hash.

Recursion across private methods is forbidden by the language
validator, so the call-graph walk terminates.
"""

from __future__ import annotations

from dataclasses import dataclass

from .ast_nodes import (
    ArrayLiteralExpr,
    AssignmentStmt,
    BinaryExpr,
    CallExpr,
    ContractNode,
    DecrementExpr,
    ExpressionStmt,
    ForStmt,
    Identifier,
    IfStmt,
    IncrementExpr,
    IndexAccessExpr,
    MemberExpr,
    MethodNode,
    PropertyAccessExpr,
    ReturnStmt,
    Statement,
    TernaryExpr,
    UnaryExpr,
    VariableDeclStmt,
)


_STATE_OUTPUT_INTRINSICS = {"addOutput", "addRawOutput"}
_DATA_OUTPUT_INTRINSICS = {"addDataOutput"}


@dataclass
class MethodEffects:
    """Effects a method has on the contract's continuation. Each flag
    is True if the effect occurs anywhere reachable from the method
    body, including transitively via private-method calls."""

    mutates_state: bool = False
    has_state_output: bool = False
    has_data_output: bool = False
    uses_preimage: bool = False

    def union(self, other: "MethodEffects") -> None:
        self.mutates_state = self.mutates_state or other.mutates_state
        self.has_state_output = self.has_state_output or other.has_state_output
        self.has_data_output = self.has_data_output or other.has_data_output
        self.uses_preimage = self.uses_preimage or other.uses_preimage


@dataclass
class ContinuationShape:
    """Continuation requirements derived from MethodEffects."""

    needs_change: bool
    needs_new_amount: bool
    is_terminal: bool


def continuation_shape_for(eff: MethodEffects) -> ContinuationShape:
    needs_change = (
        eff.mutates_state or eff.has_state_output or eff.has_data_output
    )
    # addOutput / addRawOutput already specify per-output amounts, so
    # when those are present the single-output _newAmount is redundant.
    # Otherwise the single-output continuation path needs _newAmount to
    # size the new state UTXO.
    needs_new_amount = (
        (eff.mutates_state or eff.has_data_output) and not eff.has_state_output
    )
    return ContinuationShape(
        needs_change=needs_change,
        needs_new_amount=needs_new_amount,
        is_terminal=not needs_change,
    )


def compute_side_effect_summary(contract: ContractNode) -> dict[str, MethodEffects]:
    """Classify every method on the contract.

    On-demand DFS with memoization. Returns a dict keyed by method name
    (constructor included under the key ``"constructor"``).
    """

    summary: dict[str, MethodEffects] = {}
    mutable_props: set[str] = {p.name for p in contract.properties if not p.readonly}
    private_by_name: dict[str, MethodNode] = {
        m.name: m
        for m in contract.methods
        if m.visibility != "public"
    }
    in_progress: set[str] = set()

    def classify(method_name: str, body: list[Statement]) -> MethodEffects:
        if method_name in summary:
            return summary[method_name]
        if method_name in in_progress:
            # Validation should reject recursion before we get here;
            # return empty effects defensively to avoid infinite loops.
            return MethodEffects()
        in_progress.add(method_name)
        effects = MethodEffects()
        for stmt in body:
            _collect_stmt(stmt, effects, mutable_props, private_by_name, classify)
        in_progress.discard(method_name)
        summary[method_name] = effects
        return effects

    classify("constructor", contract.constructor.body)
    for m in contract.methods:
        classify(m.name, m.body)
    return summary


def _collect_stmt(
    stmt: Statement,
    into: MethodEffects,
    mutable_props: set[str],
    private_by_name: dict[str, MethodNode],
    classify,
) -> None:
    if isinstance(stmt, AssignmentStmt):
        if isinstance(stmt.target, PropertyAccessExpr):
            if stmt.target.property in mutable_props:
                into.mutates_state = True
        if stmt.value is not None:
            _collect_expr(stmt.value, into, mutable_props, private_by_name, classify)
        return
    if isinstance(stmt, ExpressionStmt):
        if stmt.expr is not None:
            _collect_expr(stmt.expr, into, mutable_props, private_by_name, classify)
        return
    if isinstance(stmt, IfStmt):
        if stmt.condition is not None:
            _collect_expr(stmt.condition, into, mutable_props, private_by_name, classify)
        for inner in stmt.then:
            _collect_stmt(inner, into, mutable_props, private_by_name, classify)
        for inner in stmt.else_:
            _collect_stmt(inner, into, mutable_props, private_by_name, classify)
        return
    if isinstance(stmt, ForStmt):
        if stmt.update is not None:
            _collect_stmt(stmt.update, into, mutable_props, private_by_name, classify)
        for inner in stmt.body:
            _collect_stmt(inner, into, mutable_props, private_by_name, classify)
        return
    if isinstance(stmt, ReturnStmt):
        if stmt.value is not None:
            _collect_expr(stmt.value, into, mutable_props, private_by_name, classify)
        return
    if isinstance(stmt, VariableDeclStmt):
        if stmt.init is not None:
            _collect_expr(stmt.init, into, mutable_props, private_by_name, classify)
        return


def _collect_expr(
    expr,
    into: MethodEffects,
    mutable_props: set[str],
    private_by_name: dict[str, MethodNode],
    classify,
) -> None:
    if expr is None:
        return
    if isinstance(expr, (IncrementExpr, DecrementExpr)):
        if isinstance(expr.operand, PropertyAccessExpr):
            if expr.operand.property in mutable_props:
                into.mutates_state = True
        return
    if isinstance(expr, CallExpr):
        callee = expr.callee
        callee_name: str | None = None
        if isinstance(callee, PropertyAccessExpr):
            callee_name = callee.property
        elif isinstance(callee, MemberExpr):
            callee_name = callee.property

        if callee_name is not None:
            if callee_name in _STATE_OUTPUT_INTRINSICS:
                into.has_state_output = True
            if callee_name in _DATA_OUTPUT_INTRINSICS:
                into.has_data_output = True
            target = private_by_name.get(callee_name)
            if target is not None:
                into.union(classify(target.name, target.body))

        if isinstance(callee, Identifier):
            if callee.name == "checkPreimage":
                into.uses_preimage = True
            target = private_by_name.get(callee.name)
            if target is not None:
                into.union(classify(target.name, target.body))

        for arg in expr.args:
            _collect_expr(arg, into, mutable_props, private_by_name, classify)
        if not isinstance(callee, Identifier):
            _collect_expr(callee, into, mutable_props, private_by_name, classify)
        return
    if isinstance(expr, BinaryExpr):
        _collect_expr(expr.left, into, mutable_props, private_by_name, classify)
        _collect_expr(expr.right, into, mutable_props, private_by_name, classify)
        return
    if isinstance(expr, UnaryExpr):
        _collect_expr(expr.operand, into, mutable_props, private_by_name, classify)
        return
    if isinstance(expr, TernaryExpr):
        _collect_expr(expr.condition, into, mutable_props, private_by_name, classify)
        _collect_expr(expr.consequent, into, mutable_props, private_by_name, classify)
        _collect_expr(expr.alternate, into, mutable_props, private_by_name, classify)
        return
    if isinstance(expr, IndexAccessExpr):
        _collect_expr(expr.object, into, mutable_props, private_by_name, classify)
        _collect_expr(expr.index, into, mutable_props, private_by_name, classify)
        return
    if isinstance(expr, MemberExpr):
        _collect_expr(expr.object, into, mutable_props, private_by_name, classify)
        return
    if isinstance(expr, ArrayLiteralExpr):
        for el in expr.elements:
            _collect_expr(el, into, mutable_props, private_by_name, classify)
        return

File: packages/runar-compiler/src/passes/side-effect-summary.ts
/**
 * Side-effect summary pass.
 *
 * Classifies each method on a `ContractNode` by the side effects it has
 * on the contract's continuation requirements. Walks the private-method
 * call graph so that effects buried inside private helpers surface to
 * their public callers.
 *
 * Consumed by:
 *   - `04-anf-lower.ts` for auto-injecting continuation parameters
 *     (`_changePKH`, `_changeAmount`, `_newAmount`, `txPreimage`) and
 *     gating emission of the `hashOutputs` continuation assertion.
 *   - `artifact/assembler.ts` for declaring the matching ABI parameters
 *     and `isTerminal` flag.
 *
 * Both consumers must read from one source of truth so the deployed ABI
 * cannot drift from the locking script's actual parameter expectations.
 *
 * Recursion across private methods is forbidden by the language
 * validator (`02-validate.ts`), so the call-graph walk terminates.
 */
import type { ContractNode, MethodNode, Statement, Expression } from '../ir/index.js';

/**
 * Effects a method has on the contract's continuation. Each flag is
 * `true` if the effect occurs anywhere reachable from the method body,
 * including transitively via private-method calls.
 */
export interface MethodEffects {
  /** Mutates a non-readonly property (assignment or `++`/`--`). */
  mutatesState: boolean;
  /** Calls `this.addOutput(...)` or `this.addRawOutput(...)`. */
  hasStateOutput: boolean;
  /** Calls `this.addDataOutput(...)`. */
  hasDataOutput: boolean;
  /** Calls `checkPreimage(...)` (manually, outside the auto-injected one). */
  usesPreimage: boolean;
}

/** Map from method name to that method's effects. Includes the constructor under the key 'constructor'. */
export type SideEffectSummary = Map<string, MethodEffects>;

const STATE_OUTPUT_INTRINSICS = new Set(['addOutput', 'addRawOutput']);
const DATA_OUTPUT_INTRINSICS = new Set(['addDataOutput']);

function emptyEffects(): MethodEffects {
  return {
    mutatesState: false,
    hasStateOutput: false,
    hasDataOutput: false,
    usesPreimage: false,
  };
}

function unionInto(target: MethodEffects, source: MethodEffects): void {
  target.mutatesState ||= source.mutatesState;
  target.hasStateOutput ||= source.hasStateOutput;
  target.hasDataOutput ||= source.hasDataOutput;
  target.usesPreimage ||= source.usesPreimage;
}

/**
 * Compute side-effect summary for every method on the contract.
 *
 * The walk is memoized: each method is classified once, then reused if
 * referenced by another method. Visit ordering is on-demand DFS — the
 * caller does not need a topological sort.
 */
export function computeSideEffectSummary(contract: ContractNode): SideEffectSummary {
  const summary: SideEffectSummary = new Map();
  const mutablePropNames = new Set(
    contract.properties.filter(p => !p.readonly).map(p => p.name),
  );
  const privateMethodsByName = new Map<string, MethodNode>();
  for (const method of contract.methods) {
    if (method.visibility === 'private') {
      privateMethodsByName.set(method.name, method);
    }
  }

  // Memoization with an in-progress sentinel. Validation forbids
  // recursion so the sentinel should never actually be hit, but we
  // keep it as a defensive guard against stack overflow on
  // mis-validated input.
  const inProgress = new Set<string>();

  function effectsFor(methodName: string, body: Statement[]): MethodEffects {
    const cached = summary.get(methodName);
    if (cached) return cached;
    if (inProgress.has(methodName)) {
      // Treat in-progress (cyclic) calls as having no extra effects;
      // validation should have rejected the contract before we got
      // here. Returning empty avoids infinite recursion.
      return emptyEffects();
    }
    inProgress.add(methodName);
    const effects = computeBodyEffects(body);
    inProgress.delete(methodName);
    summary.set(methodName, effects);
    return effects;
  }

  function computeBodyEffects(stmts: Statement[]): MethodEffects {
    const effects = emptyEffects();
    for (const stmt of stmts) {
      collectStmt(stmt, effects);
    }
    return effects;
  }

  function collectStmt(stmt: Statement, into: MethodEffects): void {
    switch (stmt.kind) {
      case 'assignment':
        if (
          stmt.target.kind === 'property_access' &&
          mutablePropNames.has(stmt.target.property)
        ) {
          into.mutatesState = true;
        }
        // Right-hand side may still contain calls.
        collectExpr(stmt.value, into);
        return;
      case 'expression_statement':
        collectExpr(stmt.expression, into);
        return;
      case 'if_statement':
        collectExpr(stmt.condition, into);
        for (const inner of stmt.then) collectStmt(inner, into);
        if (stmt.else) for (const inner of stmt.else) collectStmt(inner, into);
        return;
      case 'for_statement':
        collectStmt(stmt.update, into);
        for (const inner of stmt.body) collectStmt(inner, into);
        return;
      case 'return_statement':
        if (stmt.value) collectExpr(stmt.value, into);
        return;
      case 'variable_decl':
        collectExpr(stmt.init, into);
        return;
      default:
        return;
    }
  }

  function collectExpr(expr: Expression, into: MethodEffects): void {
    switch (expr.kind) {
      case 'increment_expr':
      case 'decrement_expr':
        if (
          expr.operand.kind === 'property_access' &&
          mutablePropNames.has(expr.operand.property)
        ) {
          into.mutatesState = true;
        }
        return;
      case 'call_expr': {
        const callee = expr.callee;

        // this.X(...) or member.X(...) — output intrinsics or private method calls.
        if (callee.kind === 'property_access' || callee.kind === 'member_expr') {
          const name = callee.property;
          if (STATE_OUTPUT_INTRINSICS.has(name)) into.hasStateOutput = true;
          if (DATA_OUTPUT_INTRINSICS.has(name)) into.hasDataOutput = true;
          const target = privateMethodsByName.get(name);
          if (target) {
            unionInto(into, effectsFor(target.name, target.body));
          }
        }

        // Bareword calls: identifiers that resolve to private methods (Go/Rust
        // surface formats route private helpers as bare identifiers) or to
        // builtins like `checkPreimage`.
        if (callee.kind === 'identifier') {
          if (callee.name === 'checkPreimage') into.usesPreimage = true;
          const target = privateMethodsByName.get(callee.name);
          if (target) {
            unionInto(into, effectsFor(target.name, target.body));
          }
        }

        // Walk subexpressions for nested calls / property accesses.
        for (const arg of expr.args) collectExpr(arg, into);
        if (callee.kind !== 'identifier') collectExpr(callee, into);
        return;
      }
      case 'binary_expr':
        collectExpr(expr.left, into);
        collectExpr(expr.right, into);
        return;
      case 'unary_expr':
        collectExpr(expr.operand, into);
        return;
      case 'ternary_expr':
        collectExpr(expr.condition, into);
        collectExpr(expr.consequent, into);
        collectExpr(expr.alternate, into);
        return;
      case 'index_access':
        collectExpr(expr.object, into);
        collectExpr(expr.index, into);
        return;
      case 'member_expr':
        collectExpr(expr.object, into);
        return;
      case 'array_literal':
        for (const el of expr.elements) collectExpr(el, into);
        return;
      default:
        return;
    }
  }

  // Classify constructor + every method up front so callers do not
  // need to know about lazy evaluation order.
  effectsFor('constructor', contract.constructor.body);
  for (const method of contract.methods) {
    effectsFor(method.name, method.body);
  }

  return summary;
}

/**
 * Classify a method's continuation requirements based on its effects.
 *
 * `needsChange` controls injection of `_changePKH` and `_changeAmount`.
 * `needsNewAmount` controls injection of `_newAmount`. The pair maps
 * directly to ANF auto-param insertion and ABI param declaration; both
 * sites must agree for a deployed contract to be spendable.
 *
 * `isTerminal` is the inverse of `needsChange` for stateful public
 * methods — a method is terminal iff it produces no continuation.
 */
export interface ContinuationShape {
  needsChange: boolean;
  needsNewAmount: boolean;
  isTerminal: boolean;
}

export function continuationShape(effects: MethodEffects): ContinuationShape {
  const needsChange =
    effects.mutatesState || effects.hasStateOutput || effects.hasDataOutput;
  // `addOutput`/`addRawOutput` already specify per-output amounts, so
  // when those are present the single-output `_newAmount` is redundant.
  // Otherwise (mutating-only or data-only methods) the single-output
  // continuation path needs `_newAmount` to size the new state UTXO.
  const needsNewAmount =
    (effects.mutatesState || effects.hasDataOutput) && !effects.hasStateOutput;
  return {
    needsChange,
    needsNewAmount,
    isTerminal: !needsChange,
  };
}
